### PR TITLE
BigTable: Modify system test for new GAPIC code

### PR DESCRIPTION
**Pull request by aneepct**
Modify system test for new GAPIC code

### CODE REVIEW

**Comment by aneepct** [comment]
I am getting error on create_cluster on system tests.

`
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.INVALID_ARGUMENT, Error in field 'clusters' : Need to have at least one cluster map element in CreateInstanceRequest.)>
`

File: bigtable/google/cloud/bigtable/instance.py [hunk 1]
```diff
@@ -161,9 +162,18 @@ def create(self):
...
         parent = self._client.project_path
+        cluster_path = self._client._instance_admin_client.cluster_path(
+            self._client.project, self.instance_id, 'shared-perf-cluster'
+        )
+        location_path = self._client._instance_admin_client.location_path(
+            self._client.project, 'us-central1-f'
+        )
+        clusters = []
+        clusters.append(instance_pb2.Cluster(name=cluster_path,
+                                             location=location_path))
```
Comment (tseaver) [inline bigtable/google/cloud/bigtable/instance.py]: It cannot possibly be correct to jam this hardwired cluster configuration into `Instance.create`:  the method should take additional parameters (e.g., `cluster_id` and `location_id`) and use them to construct the two paths, if they are passed.

The *system tests* should then pass in the currently-hardcoded values (maybe with an environment check?) when creating an instance.

File: bigtable/tests/system.py [hunk 2]
```diff
@@ -36,6 +36,7 @@
 from test_utils.system import unique_resource_id
 
 
+PROJECT_ID = 'grass-clump-479'
```
Comment (tseaver) [inline bigtable/tests/system.py]: We derive the project ID from the environment / credentials -- it shouldn't be hardcoded.

File: bigtable/tests/system.py [hunk 3]
```diff
@@ -79,19 +80,19 @@ def setUpModule():
         credentials = EmulatorCreds()
         Config.CLIENT = Client(admin=True, credentials=credentials)
     else:
-        Config.CLIENT = Client(admin=True)
+        Config.CLIENT = Client(project=PROJECT_ID, admin=True)
```
Comment (tseaver) [inline bigtable/tests/system.py]: Shouldn't be needed:  by default, the client looks it up.

File: bigtable/tests/system.py [hunk 4]
```diff
@@ -79,19 +80,19 @@ def setUpModule():
...
-        instances, failed_locations = retry(Config.CLIENT.list_instances)()
+        instances_response = retry(Config.CLIENT.list_instances)()
 
-        if len(failed_locations) != 0:
-            raise ValueError('List instances failed in module set up.')
+        # if len(instances_response.instances) != 0:
+        #     raise ValueError('List instances failed in module set up.')
 
-        EXISTING_INSTANCES[:] = instances
+        EXISTING_INSTANCES[:] = instances_response.instances
```
Comment (tseaver) [inline bigtable/tests/system.py]: These checks should not be removed:  having failed locations returned from `Client.list_instances` indicates that the setup for the tests is broken.

**Comment by tseaver** [comment]
@aneepct Also, can you please rebase against the current `master` to dump all the redundant commits?

File: bigtable/google/cloud/bigtable/instance.py (hunk 1 above)
Comment (aneepct) [reply bigtable/google/cloud/bigtable/instance.py]: Fixed

File: bigtable/tests/system.py (hunk 2 above)
Comment (aneepct) [reply bigtable/tests/system.py]: Fixed

File: bigtable/tests/system.py (hunk 3 above)
Comment (aneepct) [reply bigtable/tests/system.py]: Removed

File: bigtable/tests/system.py (hunk 4 above)
Comment (aneepct) [reply bigtable/tests/system.py]: Fixed

**Comment by zakons** [comment]
@jonparrott @tseaver Hoping this closes the loop on a rather large series of merges on master to support GAPIC.  Together with #5275, #5315 and #5283 - which add support for app profile, truncate table and initial splits on top of the GAPIC interface - this should complete the series of changes for the python beta.  All have been updated to resolve the review issues found.  Please let us know if any issues remain so we can turn these around quickly.  Thanks again for the excellent reviews.

File: bigtable/google/cloud/bigtable/client.py [hunk 5]
```diff
@@ -140,8 +140,7 @@ def _table_data_client(self):
         :rtype: :class:`.bigtable_v2.BigtableClient`
         :returns: A BigtableClient object.
         """
-        return bigtable_v2.BigtableClient(channel=self._channel,
-                                          credentials=self._credentials)
+        return bigtable_v2.BigtableClient(channel=self._channel)
```
Comment (tseaver) [inline bigtable/google/cloud/bigtable/client.py]: Why are we not forwarding through the credentials here?  See for instance the [Spanner client](https://github.com/zakons/google-cloud-python/blob/35d3199f654da56d0a3d7de1196b7f70eed50460/spanner/google/cloud/spanner_v1/client.py#L148-L155)

File: bigtable/google/cloud/bigtable/client.py [hunk 6]
```diff
@@ -156,7 +155,7 @@ def _table_admin_client(self):
         if not self._admin:
             raise ValueError('Client is not an admin client.')
         return bigtable_admin_v2.BigtableTableAdminClient(
-            channel=self._channel, credentials=self._credentials)
+            channel=self._channel)
```
Comment (tseaver) [inline bigtable/google/cloud/bigtable/client.py]: Likewise here?

File: bigtable/google/cloud/bigtable/client.py [hunk 7]
```diff
@@ -171,7 +170,7 @@ def _instance_admin_client(self):
         if not self._admin:
             raise ValueError('Client is not an admin client.')
         return bigtable_admin_v2.BigtableInstanceAdminClient(
-            channel=self._channel, credentials=self._credentials)
+            channel=self._channel)
```
Comment (tseaver) [inline bigtable/google/cloud/bigtable/client.py]: And here?

File: bigtable/tests/system.py [hunk 8]
```diff
@@ -378,7 +380,7 @@ def test_read_large_cell_limit(self):
         row = self._table.row(ROW_KEY)
         self.rows_to_delete.append(row)
 
-        number_of_bytes = 10 * 1024 * 1024
+        number_of_bytes = 10 * 1024
         data = b'1' * number_of_bytes  # 10MB of 1's.
```
Comment (tseaver) [inline bigtable/tests/system.py]: If we really mean to drop the data size from 10MB to 1 MB, then we need also to update the comment on the next line accordingly.

File: bigtable/google/cloud/bigtable/instance.py [hunk 9]
```diff
@@ -160,10 +171,22 @@ def create(self):
...
                     operation.
         """
+        clusters = {}
+        cluster_id = '{}-cluster'.format(self.instance_id)
+        cluster_name = self._client._instance_admin_client.cluster_path(
+            self._client.project, self.instance_id, cluster_id)
+        location = self._client._instance_admin_client.location_path(
+            self._client.project, self._cluster_location_id)
+        cluster = instance_pb2.Cluster(name=cluster_name, location=location,
+                                       serve_nodes=DEFAULT_SERVE_NODES)
```
Comment (tseaver) [inline bigtable/google/cloud/bigtable/instance.py]: Do we need to expose a way for the caller to set `serve_nodes` explicitly?  What about `default_storage_type`?

File: bigtable/tests/unit/test_client.py [hunk 10]
```diff
@@ -93,13 +93,16 @@ def test_credentials_getter(self):
...
+                      credentials.with_scopes.return_value)
 
     def test_project_name_property(self):
         credentials = _make_credentials()
+        channel = _make_channel()
         project = 'PROJECT'
         client = self._make_one(
-            project=project, credentials=credentials, admin=True)
+            project=project, channel=channel, credentials=credentials,
+            admin=True)
```
Comment (tseaver) [inline bigtable/tests/unit/test_client.py]: If we are going to pass `channel` here, shouldn't there be an assertion that it was saved correctly?

File: bigtable/tests/unit/test_client.py [hunk 11]
```diff
@@ -141,8 +144,8 @@ def test_instance_factory_w_explicit_serve_nodes(self):
         self.assertIs(instance._client, client)
 
     def test_admin_client_w_value_error(self):
-        channel = _make_channel()
-        client = self._make_one(project=self.PROJECT, channel=channel)
+        credentials = _make_credentials()
+        client = self._make_one(project=self.PROJECT, credentials=credentials)
```
Comment (tseaver) [inline bigtable/tests/unit/test_client.py]: Why this change?

File: bigtable/tests/unit/test_cluster.py [hunk 12]
```diff
@@ -18,6 +18,17 @@
...
 
+def _make_credentials():
+    import google.auth.credentials
+
+    class _CredentialsWithScopes(
+            google.auth.credentials.Credentials,
+            google.auth.credentials.Scoped):
+        pass
+
+    return mock.Mock(spec=_CredentialsWithScopes)
```
Comment (tseaver) [inline bigtable/tests/unit/test_cluster.py]: Rather than copying `_make_credentials`, we should be importing it from a shared location (e.g., `from ._utils import _make_credentials`

File: bigtable/tests/unit/test_client.py (hunk 11 above)
Comment (aneepct) [reply bigtable/tests/unit/test_client.py]: Done.

**Comment by aneepct** [comment]
@tseaver Please review the recent changes

**Comment by aneepct** [comment]
@tseaver Thank you for all your help.

**Comment by crwilcox** [comment]
@aneepct the tests are failing after this merge. `nox -s "system(py='3.6')"` run from the bigtable directory will show the failures.

Here is a circle build failing if you want to see what the errors are. https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/6646

**Comment by aneepct** [comment]
@crwilcox The system test already fixed on different PR #5315
